### PR TITLE
Hook up event emitters between napari TransformChain and vispy

### PR DESCRIPTION
**Pull request by GenevieveBuckley**
# Description
This PR hooks up event emitters from the napari TransformChain class with vispy.

### CODE REVIEW

File: napari/_vispy/vispy_transforms.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            translate = np.flip(t.translate)  # vispy ordered array
+            scale = np.flip(t.scale)  # vispy ordered array
+            vispy_transforms.append(
+                STTransform(scale=scale, translate=translate)
+            )
+        self.transform_chain = ChainTransform(vispy_transforms)
+        return self.transform_chain
+
+    def _calculate_attributes(self):
+        """Simplified transformation matrix, scale, and transle attributes.
```
Comment (GenevieveBuckley) [inline napari/_vispy/vispy_transforms.py:98]: All this attribute stuff is so I can try switching out the STTransform in `_master_transform` for an instance of our new VispyChainTransform class. I want to try running the test suite before making more changes (but I expect we will tidy this up later on).

File: napari/_vispy/vispy_transforms.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        self.transform_chain.transforms.insert(transform_index, transform)
+        self._calculate_attributes()
+
+    def _remove(self, event):
+        """Remove vispy transform at index `event.index`."""
+        self.transform_chain.transforms.pop(event.index)
+        self._calculate_attributes()
+
+    def _reorder(self, event=None):
+        raise NotImplementedError()
```
Comment (GenevieveBuckley) [inline napari/_vispy/vispy_transforms.py:137]: I'm not entirely sure if we need this?

File: napari/layers/transforms.py [hunk 3]
```diff
@@ -24,6 +24,8 @@ def __init__(self, func=tz.identity, inverse=None, name=None):
         self.name = name
         if func is tz.identity:
             self._inverse_func = tz.identity
+        self.translate = None
+        self.scale = None
```
Comment (GenevieveBuckley) [inline napari/layers/transforms.py:28]: This means we don't run into AttributeErrors later on, and keeps the downstream code a little more elegant.

File: napari/_vispy/vispy_transforms.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        self.transform_chain.events.reordered.connect(self._reorder)
+
+    def _vispy_transform_chain(self):
+        """Builds vispy ChainTransform from a napari TransformChain."""
+        vispy_transforms = []
+        for t in self.napari_transform_chain:
+            # From napari format (t, z, y, x)
+            # to the vispy format (x, y, z, ?)
+            translate = np.flip(t.translate)  # vispy ordered array
+            scale = np.flip(t.scale)  # vispy ordered array
```
Comment (GenevieveBuckley) [inline napari/_vispy/vispy_transforms.py]: Is my understanding here correct? I think the napari transforms have the format (T,Z,Y,X), and the vispy transforms have the format (X,Y,Z,T). Additionally napari transform scale and translate arrays can have fewer than 4 elements, eg if you're just specifying a translation in x and y.

**Comment by GenevieveBuckley** [comment]
My next step here is try integrating this with the vispy base layer class.

@jni - [I have a question for you here](https://github.com/jni/napari/pull/5#discussion_r381812656)., and feel free to let me know if you think I'm going in the wrong direction more generally. Otherwise,  this is still very rough, so don't spend too much time looking over it just yet.

@sofroniewn - in qt_layerlist.py you have [some keypress functions](https://github.com/napari/napari/blob/bd538488c9a6f9d6721cec8268610a1ea5b0c7d0/napari/_qt/qt_layerlist.py#L189-L197) that just do event.ignore(). Does it cause a problem to have an event emitting but not being explicitly ignored?
```python
def keyPressEvent(self, event):
    event.ignore()
```

I'm not sure if I will also need to have a bunch of event functions that do nothing except `event.ignore()`. It seems like that would be silly and uncecessary in this context, but I don't have a full understanding of the logic for it in qt_layerlist.py

**Comment by sofroniewn** [comment]
> @sofroniewn - in qt_layerlist.py you have some keypress functions that just do event.ignore().

The `event.ignore` in the qt_layerlist allows the event to pass through a parent widget to its child widget without doing anything, otherwise the parent catches the event and it doesn't get passed on. This was needed there, but shouldn't be needed in your context here, where I think you're just emitting events that no one is connected to yet.

**Comment by GenevieveBuckley** [comment]
I'm sorry, I merged master into my working branch and now it's impossible for you to easily see what I've done. I think the simplest way forward is if you merge master into your own world-coords branch (unless there's another way you'd like me to fix it?)

* I've now made a very rough start on replacing the STTransform in the vispy base layer class with the new VispyChainTransform class.
* Then, there were errors in the tests related to image pyramids. I made some changes here (not great changes, I've sidestepped the error messages but introduced other problems). 
* All the tests pass right now (but again, there are problems the tests aren't checking for). An example of this kind of problem is the output I get from running `example/add_pyramid.py`, see this screenshot:

pyramid_problems

The same script from the napari master branch produces a much better looking display of the image pyramid. This is what we want to have:
pyramid_expected

**Comment by sofroniewn** [comment]
@GenevieveBuckley one the merge stuff is sorted out I can try and help with the image pyramid stuff, there definitely might be some weird things there that are my fault

**Comment by GenevieveBuckley** [comment]
> @GenevieveBuckley one the merge stuff is sorted out I can try and help with the image pyramid stuff, there definitely might be some weird things there that are my fault

There are actually a few things in master we might need or want (eg: Talley's changed event https://github.com/napari/napari/pull/982)

**Comment by GenevieveBuckley** [comment]
Ok, this version might be easier for you to look at @sofroniewn https://github.com/jni/napari/pull/6
I can't figure out why the behavior is different - at first I thought it was flipping/not flipping the scale or translation array, but I'm not sure that's actually the root problem...

All of Juan's changes on this branch are in napari/layers/transforms.py, so there aren't any hiding that you might miss.

**Comment by GenevieveBuckley** [comment]
Closing in favour of https://github.com/jni/napari/pull/6

**Pull request closed without merging**